hugegraph-server/hugegraph-dist/src/main/java/org/apache/hugegraph/dist/MemoryMonitor.java

```diff
@@ -0,0 +1,119 @@
+
+public class MemoryMonitor {
+
+    private static final Logger LOG = Log.logger(MemoryMonitor.class);
+    private final double MEMORY_MONITOR_THRESHOLD;
+    private final int MEMORY_MONITOR_PERIOD;
+    private final ScheduledExecutorService scheduler;
+
+    public MemoryMonitor(String restServerConf) {
+        HugeConfig restServerConfig = new HugeConfig(restServerConf);
+        MEMORY_MONITOR_THRESHOLD =
+                restServerConfig.get(ServerOptions.JVM_MEMORY_MONITOR_THRESHOLD);
+        MEMORY_MONITOR_PERIOD =
+                restServerConfig.get(ServerOptions.JVM_MEMORY_MONITOR_PERIOD);
+        scheduler = Executors.newScheduledThreadPool(1);
+    }
+
+    private void manageMemory() {
+        double memoryUsagePercentage = getMemoryUsagePercentage();
+
+        if (memoryUsagePercentage > MEMORY_MONITOR_THRESHOLD) {
+            LOG.info("JVM memory usage is '{}', exceeding the threshold of '{}'.",
+                     memoryUsagePercentage, MEMORY_MONITOR_THRESHOLD);
+            System.gc();
+            LOG.info("Trigger System.gc()");
+
+            double doubleCheckUsage = getMemoryUsagePercentage();
+            if (doubleCheckUsage > MEMORY_MONITOR_THRESHOLD) {
+                Thread targetThread = getHighestMemoryThread();
+                if (targetThread != null) {
+                    LOG.info("JVM memory usage is '{}', exceeding the threshold of '{}'.",
+                             doubleCheckUsage, MEMORY_MONITOR_THRESHOLD);
+                    targetThread.interrupt();
+                    LOG.info("Send interrupt to '{}' thread",
+                             targetThread.getName());
+                }
+            }
+        }
+    }
+
+    private double getMemoryUsagePercentage() {
+        MemoryUsage heapMemoryUsage =
+                ManagementFactory.getMemoryMXBean().getHeapMemoryUsage();
+        return (double) heapMemoryUsage.getUsed() / heapMemoryUsage.getMax();
+    }
+
+    private Thread getHighestMemoryThread() {
+        long highestMemory = 0;
+        Thread highestThread = null;
+
+        ThreadMXBean threadMXBean = (ThreadMXBean) ManagementFactory.getThreadMXBean();
+
+        for (Thread thread : Thread.getAllStackTraces().keySet()) {
+            if (thread.getState() != Thread.State.RUNNABLE ||
+                thread.getName() == null ||
+                !thread.getName().startsWith("grizzly-http-server-")) {
+                continue;
+            }
+
+            long threadMemory =
+                    threadMXBean.getThreadAllocatedBytes(thread.getId());
+            if (threadMemory > highestMemory) {
+                highestMemory = threadMemory;
+                highestThread = thread;
+            }
+        }
+
+        return highestThread;
+    }
+
+    public void start() {
+        if (MEMORY_MONITOR_THRESHOLD >= 1.0) {
+            return;
+        }
+        Runnable task = this::manageMemory;
+        scheduler.scheduleAtFixedRate(task, 0, MEMORY_MONITOR_PERIOD,
```

Comment:
maybe submit the task to the thread pool directly in the main thread, without creating a new thread to submit the scheduled task?

Reply:
I don't get your idea clearly. Now I have created a thread pool and submitted a task into it. Do you mean not to create a new thread pool in this class?

Reply:
I was wrong here. There is no problem with the implementation, but it can be more elegant. such as `scheduler.scheduleAtFixedRate(this::manageMemory, ...)